hbase_sql/src/main/java/com/splicemachine/derby/stream/spark/NativeSparkDataSet.java

```diff
@@ -205,17 +211,40 @@ public void unpersistIt() {
      */
     @SuppressWarnings({ "unchecked", "rawtypes" })
     @Override
-    public DataSet<V> distinct(OperationContext context) {
+    public DataSet<V> distinct(OperationContext context) throws StandardException {
         return distinct("Remove Duplicates", false, context, false, null);
     }
 
     @SuppressWarnings({ "unchecked", "rawtypes" })
     @Override
-    public DataSet<V> distinct(String name, boolean isLast, OperationContext context, boolean pushScope, String scopeDetail) {
+    public DataSet<V> distinct(String name, boolean isLast, OperationContext context, boolean pushScope, String scopeDetail) throws StandardException {
+        String varcharDB2CompatibilityModeString =
+        PropertyUtil.getCachedDatabaseProperty(context.getActivation().getLanguageConnectionContext(), Property.SPLICE_DB2_VARCHAR_COMPATIBLE);
+        boolean varcharDB2CompatibilityMode = Boolean.parseBoolean(varcharDB2CompatibilityModeString);
         pushScopeIfNeeded(context, pushScope, scopeDetail);
         try {
-            Dataset<Row> result = dataset.distinct();
+            // Do not replace string columns using rtrimmed columns but add them as extra columns.
+            // After distinct, for each group, we need to return a row originally in that group, not
+            // a row with rtrimmed values.
+            List<String> compColNames = new ArrayList<>();
+            if (varcharDB2CompatibilityMode) {
+                List<String> strColNames = new ArrayList<>();
+                Tuple2<String, String>[] colTypes = dataset.dtypes();
+                for (Tuple2<String, String> colType : colTypes) {
+                    if (colType._2().equals("StringType")) {
+                        strColNames.add(colType._1());
+                    } else {
+                        compColNames.add(colType._1());
+                    }
+                }
+                for (String colName : strColNames) {
+                    String newColName = colName+"_rtrimmed";
+                    dataset = dataset.withColumn(newColName, rtrim(col(colName)));
+                    compColNames.add(newColName);
+                }
+            }
 
+            Dataset<Row> result = dataset.dropDuplicates(compColNames.toArray(new String[0]));
```

Comment:
The fix looks good, but when varcharDB2CompatibilityMode is false I think we need to return dataset.distinct().

Reply:
Yes, that's right. For clarity, I still changed it to use `distinct()` when `compColNames` is empty, partly also because I was not able to verify it based on Spark's source code. Their `dropDuplicates()` implementation (no args) passes `df.columns` in.